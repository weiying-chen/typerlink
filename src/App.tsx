import React from 'react';
import { useState, useEffect, useRef, ChangeEvent } from 'react';
import {
	getPrevious,
	getNext,
	findElementsByText,
	addHighlight,
	removeHighlight,
	addSelectedClass,
} from './utils';
// import './style.css';

type ContentEditableElement = HTMLElement & {
	contentEditable: 'true' | 'false';
	innerHTML: string;
};

function App() {
	const [inputValue, setInputValue] = useState('');
	const [elements, setElements] = useState<HTMLElement[]>([]);
	const [selectedElement, setSelectedElement] = useState<HTMLElement | null>(
		null
	);
	// TODO: Maybe merge this with `inputValue`?
	const [isInputFocused, setIsInputFocused] = useState(false);

	// This is needed because `selectedElement` only the initial state inside `handleDocumentKeyDown`.
	const elementsRef = useRef(elements);
	const selectedElementRef = useRef(selectedElement);
	const inputRef = useRef<HTMLInputElement | null>(null);

	selectedElementRef.current = selectedElement;
	elementsRef.current = elements;

	function handleInputChange(event: ChangeEvent<HTMLInputElement>) {
		// The default `input` behavior already ignores Ctrl commands.
		const selectors = 'a, button';
		const { value } = event.target as HTMLInputElement;

		// TODO: Rename "found" to something else?
		const foundElements = findElementsByText(selectors, value);
		const foundSelectedElement = foundElements[0] || null;

		// These functions can't be pure, because they have to access the original HTML elements
		if (elements.length) removeHighlight(elements);
		if (foundElements.length) addHighlight(foundElements, value);

		// No need to remove `selected` from the previous element because `removeHighlights` is removing `<marks>`.
		if (foundSelectedElement) {
			addSelectedClass(foundSelectedElement);
			foundSelectedElement.scrollIntoView({
				block: 'center',
				behavior: 'auto',
			});
		}

		setInputValue(value);
		setElements(foundElements);
		setSelectedElement(foundSelectedElement);
	}

	function handleInputBlur(event: ChangeEvent<HTMLInputElement>) {
		setIsInputFocused(false);
		setInputValue('');

		// Highlight won't be removed if `elements` are set to `[]`.
		if (elementsRef.current.length) removeHighlight(elementsRef.current);

		setElements([]);

		// This will also set `selectedElementRef.current to `null`.
		setSelectedElement(null);
	}

	function isInInput(event: KeyboardEvent) {
		const target = event.target as HTMLElement;

		return (
			target.nodeName === 'INPUT' ||
			target.nodeName === 'TEXTAREA' ||
			target.isContentEditable
		);
	}

	function isCommand(event: KeyboardEvent) {
		// TODO: maybe check for non-alphanumeric
		return (
			event.ctrlKey ||
			event.key === 'Enter' ||
			event.key === 'Escape' ||
			event.key === '/'
		);
	}

	function handleDocumentKeyDown(event: any) {
		if (!isCommand(event)) return;

		// TODO: Make this DRY.
		// And maybe there should only be a function inside each key event
		if (event.ctrlKey && event.key === ']') {
			const foundSelectedElement = getNext(
				elementsRef.current,
				selectedElementRef.current
			);

			selectedElementRef.current
				?.querySelector('mark')
				?.classList.remove('selected');

			if (foundSelectedElement) {
				addSelectedClass(foundSelectedElement);
				foundSelectedElement.scrollIntoView({
					block: 'center',
					behavior: 'auto',
				});
			}

			setSelectedElement(foundSelectedElement);
		}

		if (event.ctrlKey && event.key === '[') {
			const foundSelectedElement = getPrevious(
				elementsRef.current,
				selectedElementRef.current
			);

			selectedElementRef.current
				?.querySelector('mark')
				?.classList.remove('selected');

			if (foundSelectedElement) {
				addSelectedClass(foundSelectedElement);
				foundSelectedElement.scrollIntoView({
					block: 'center',
					behavior: 'auto',
				});
			}

			setSelectedElement(foundSelectedElement);
		}

		if (event.key === 'Enter') {
			if (!selectedElementRef.current) return;

			selectedElementRef.current.click();
			setInputValue('');

			// Highlight won't be removed if `elements` are set to `[]`.
			if (elementsRef.current.length)
				removeHighlight(elementsRef.current);

			setElements([]);

			// This will also set `selectedElementRef.current to `null`.
			setSelectedElement(null);
		}

		if (event.key === 'Escape') {
			(
				document.activeElement as
					| HTMLInputElement
					| HTMLTextAreaElement
					| ContentEditableElement
			)?.blur();
		}

		if (event.key === '/' && !isInInput(event)) {
			// Prevents `/` from being input.
			event.preventDefault();
			setIsInputFocused(true);
		}
	}

	useEffect(() => {
		// A React event handler can't be used, because it can't access `document`.
		document.addEventListener('keydown', handleDocumentKeyDown);

		return () => {
			document.removeEventListener('keydown', handleDocumentKeyDown);
		};
	}, []);

	useEffect(() => {
		inputRef.current?.focus();
	}, [isInputFocused]);
	
	return (
		<div id="keys" style={{ display: isInputFocused ? 'flex' : 'none' }}>
			<input
				type="text"
				ref={inputRef}
				onChange={handleInputChange}
				onBlur={handleInputBlur}
				value={inputValue}
			/>
			<span id="count">
				{/* TODO: Turn this into a variable */}
				{selectedElement ? elements.indexOf(selectedElement) + 1 : 0}/
				{elements.length}
			</span>
		</div>
	);
}

export default App