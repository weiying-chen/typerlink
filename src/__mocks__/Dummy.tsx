import React from 'react';

function Dummy () {
	return (
		<div id="dummy">
			<button>Button</button>
		</div>
	);
};

export default Dummy