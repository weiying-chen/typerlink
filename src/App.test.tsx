import '@testing-library/jest-dom';
import { render, screen, fireEvent, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import React from 'react';
import App from './App';
import Dummy from './__mocks__/Dummy';

// Otherwise, the `scrollIntoView` called from `App` will throw an error.
Element.prototype.scrollIntoView = jest.fn();

describe('`input`', () => {
	test('is not visible on page load', () => {
		render(<App />);

		const input = screen.getByRole('textbox', { hidden: true });

		expect(input).not.toBeVisible();
	});

	test('is visible after pressing `/`', () => {
		render(<App />);
		fireEvent.keyDown(document, { key: '/' });

		const input = screen.getByRole('textbox');

		expect(input).toBeVisible();
	});
})

describe('`button`', () => {
	test('has been highlighted`', async () => {
		const { container } = render(
			<>
				<Dummy />
				<App />
			</>
		);

		fireEvent.keyDown(document, { key: '/' });

		const input = screen.getByRole('textbox');

		await userEvent.type(input, 'B');

		const button = screen.getByRole('button');

		expect(container).toMatchSnapshot();
		expect(button).toBeVisible();

		// TODO: Ask ChatGPT
		// 1. the best way to test <button><mark>B</mark></button> 
		// 2. If jest.mock() should be used
		// 3. If can change the `fireEvent`s to `userEvent`s. 
	});
});
