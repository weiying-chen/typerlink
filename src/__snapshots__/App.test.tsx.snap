// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Test test test\` 1`] = `
<div>
  <div
    id="dummy"
  >
    <button>
      <mark
        class="selected"
      >
        B
      </mark>
      utton
    </button>
  </div>
  <div
    id="keys"
    style="display: flex;"
  >
    <input
      type="text"
      value="B"
    />
    <span
      id="count"
    >
      1
      /
      1
    </span>
  </div>
</div>
`;
